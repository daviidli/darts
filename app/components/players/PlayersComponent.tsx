import React from 'react';
import {
	List,
	ListItem,
	TextField,
	ListItemSecondaryAction,
	IconButton,
	Button
} from '@material-ui/core';
import { MdDelete } from 'react-icons/md';
import { playerType } from '../../reducers/types';

export type Props = {
	players: playerType[];
	addPlayer: (player: string) => {};
	removePlayer: (index: number) => {};
	changePlayer: (players: string, index: number) => {};
};

const PlayersComponent = (props: Props) => {
	const { players, addPlayer, removePlayer, changePlayer } = props;
	return (
		<>
			<List>
				{players.map((player, i) => (
					<ListItem key={`player-${player.id}`}>
						<TextField
							label={`Player ${i + 1}`}
							value={player.name}
							onChange={e => changePlayer(e.target.value, i)}
							color="secondary"
						/>
						<ListItemSecondaryAction>
							<IconButton
								edge="end"
								onClick={() => removePlayer(i)}
							>
								<MdDelete />
							</IconButton>
						</ListItemSecondaryAction>
					</ListItem>
				))}
			</List>
			<Button
				variant="contained"
				color="secondary"
				onClick={() => addPlayer(`player ${players.length + 1}`)}
			>
				Add Player
			</Button>
		</>
	);
};

export default PlayersComponent;
