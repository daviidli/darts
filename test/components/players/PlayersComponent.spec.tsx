import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Button, TextField, IconButton } from '@material-ui/core';
import PlayersComponent, {
	Props
} from '../../../app/components/players/PlayersComponent';

Enzyme.configure({ adapter: new Adapter() });

const initialProps: Props = {
	players: [{ id: '1', name: 'player1' }],
	addPlayer: () => ({}),
	removePlayer: () => ({}),
	changePlayer: () => ({})
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<PlayersComponent {...props} />);
	return {
		wrapper,
		button: wrapper.find(Button),
		iconButton: wrapper.find(IconButton),
		input: wrapper.find(TextField)
	};
};

describe('Players component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot for 2 players', () => {
		const props: Props = { ...initialProps };
		props.players = [
			{
				id: '1',
				name: 'player1'
			},
			{
				id: '2',
				name: 'player2'
			}
		];
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should add new player', () => {
		const addPlayerMock = spy();
		const props = { ...initialProps };
		props.addPlayer = addPlayerMock;
		const { button } = setup(props);
		expect(button.text()).toMatch('Add Player');
		button.simulate('click');
		expect(addPlayerMock.calledOnce).toBeTruthy();
		expect(addPlayerMock.calledWith('player 2')).toBeTruthy();
	});

	it('should delete a player', () => {
		const removePlayerMock = spy();
		const props = { ...initialProps };
		props.removePlayer = removePlayerMock;
		const { iconButton } = setup(props);
		iconButton.simulate('click');
		expect(removePlayerMock.calledOnce).toBeTruthy();
		expect(removePlayerMock.calledWith(0)).toBeTruthy();
	});

	it("should edit a player's name", () => {
		const changePlayerMock = spy();
		const props = { ...initialProps };
		props.changePlayer = changePlayerMock;
		const { input } = setup(props);
		input.simulate('change', { target: { value: 'player12' } });
		expect(changePlayerMock.calledOnce).toBeTruthy();
		expect(
			changePlayerMock.firstCall.calledWith('player12', 0)
		).toBeTruthy();
	});
});
