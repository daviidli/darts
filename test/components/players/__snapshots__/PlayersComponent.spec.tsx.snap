// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players component should match snapshot 1`] = `
Array [
  <ul
    className="MuiList-root MuiList-padding"
  >
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
        disabled={false}
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
          >
            Player 1
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="player1"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
        </button>
      </div>
    </li>
  </ul>,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Add Player
    </span>
  </button>,
]
`;

exports[`Players component should match snapshot for 2 players 1`] = `
Array [
  <ul
    className="MuiList-root MuiList-padding"
  >
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
        disabled={false}
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Player 1
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="player1"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </li>
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
        disabled={false}
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Player 2
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="player2"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </li>
  </ul>,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Add Player
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>,
]
`;
